src/modules/iotjs_module_mqtt.c

```diff
@@ -0,0 +1,719 @@
+  iotjs_bufferwrap_t *topic_payload = iotjs_bufferwrap_from_jbuffer(jtopic);
+
+  uint8_t packet_identifier_lsb = 0;
+  uint8_t packet_identifier_msb = 0;
+
+  // IF NOT QOS
+  if (!jerry_value_is_undefined(jqos) &&
+      !jerry_value_is_undefined(jpacket_id) &&
+      jerry_value_is_number(jpacket_id)) {
+    uint16_t packet_identifier = jerry_get_number_value(jpacket_id);
+    packet_identifier_msb = (uint8_t)(packet_identifier >> 8);
+    packet_identifier_lsb = (uint8_t)(packet_identifier & 0x00FF);
+  }
+
+  size_t payload_len = message_payload->length + IOTJS_MQTT_LSB_MSB_SIZE;
+  size_t variable_header_len = topic_payload->length + IOTJS_MQTT_LSB_MSB_SIZE +
+                               sizeof(packet_identifier_msb) +
+                               sizeof(packet_identifier_lsb);
+  uint32_t remaining_length = payload_len + variable_header_len;
+  size_t full_len = sizeof(header_byte) +
+                    get_remaining_length_size(remaining_length) +
+                    variable_header_len + payload_len;
+
+  jerry_value_t jbuff = iotjs_bufferwrap_create_buffer(full_len);
+  iotjs_bufferwrap_t *buffer_wrap = iotjs_bufferwrap_from_jbuffer(jbuff);
+
+  uint8_t *buff_ptr = (uint8_t *)buffer_wrap->buffer;
+
+  *buff_ptr++ = header_byte;
+  buff_ptr = iotjs_encode_remaining_length(remaining_length, buff_ptr);
+  buff_ptr = iotjs_mqtt_string_serialize(buff_ptr, topic_payload);
+  *buff_ptr++ = packet_identifier_msb;
+  *buff_ptr++ = packet_identifier_lsb;
+
+  buff_ptr = iotjs_mqtt_string_serialize(buff_ptr, message_payload);
+
+  jerry_release_value(jmessage);
+  jerry_release_value(jtopic);
+  jerry_release_value(jretain);
+
+  return jbuff;
+}
+
+
+JS_FUNCTION(MqttHandle) {
+  DJS_CHECK_THIS();
+  DJS_CHECK_ARGS(2, object, object);
+
+  jerry_value_t jsref = JS_GET_ARG(0, object);
+  jerry_value_t jparam = JS_GET_ARG(1, object);
+
+  iotjs_bufferwrap_t *buffer_wrap = iotjs_bufferwrap_from_jbuffer(jparam);
+  char *buffer = buffer_wrap->buffer;
+
+  char first_byte = buffer[0];
+  char packet_type = (first_byte >> 4) & 0x0F;
+
+  size_t offset = 1;
+  uint32_t remaining_length = iotjs_decode_remaining_length(buffer, &offset);
+
+  switch (packet_type) {
+    case CONNACK: {
+      if (remaining_length != 2) {
+        return JS_CREATE_ERROR(COMMON, "MQTT: CONNACK packet is corrupted");
+      }
+
+      uint8_t return_code = (uint8_t) buffer[++offset];
+
+      if (return_code != 0) {
+        return mqtt_client_connack_error(return_code);
+      }
+
+      jerry_value_t fn =
+          iotjs_jval_get_property(jsref, IOTJS_MAGIC_STRING_ONCONNECT);
+      iotjs_make_callback(fn, jsref, iotjs_jargs_get_empty());
+      jerry_release_value(fn);
+      break;
+    }
+    case PUBLISH: {
+      MQTTHeader header = { 0 };
+      header.bits.type = PUBLISH;
+      header.bits.dup = first_byte & 0x08;
+      header.bits.qos = (first_byte & 0x06) >> 1;
+      header.bits.retain = first_byte & 0x01;
```

Comment:
I think more checks would be useful here. E.g. jerry offers utf8 validating functions.